src/Cast.hpp

```diff
@@ -0,0 +1,138 @@
+// TODO make base type, i.e. From = ours
+
+template<class FromType>
+struct Cast {
+    // for the default case type ToType = From
+    typedef FromType ToType;
+
+    // disallow to call non-specialized version
+    static ToType from(const FromType& bw) = delete;
+    static FromType to(const ToType& rs) = delete;
+};
+
+//_______________________________________________________//
+// shortcut function to make Cast usage more lightweight //
+
+// BW - types that come from BWAPI
+// RS - types that come from outside, e.g. Rust
+
+template<class BW>
+typename Cast<BW>::ToType from_bw(const BW& bw) {
+    return Cast<BW>::from(bw);
+};
+
+// TODO no
+template<class RS>
+RS to_bw(const typename Cast<RS>::ToType& rs) {
+    return Cast<RS>::to(rs);
+};
+
+//_______________________________//
+// Cast specializations go here  //
+
+template<>
+struct Cast<BWAPI::Position> {
+    typedef Position ToType;
+    inline static Position from(const BWAPI::Position& bw) {
+        return Position {bw.x, bw.y};
+    }
+    inline static BWAPI::Position to(const Position& rs) {
+        return BWAPI::Position {rs.x, rs.y};
+    }
+};
+
+
+template<>
+struct Cast<BWAPI::UnitType> {
+    typedef UnitType ToType;
+    inline static UnitType from(const BWAPI::UnitType& bw) {
+        return UnitType {bw.getID()};
+    }
+    inline static BWAPI::UnitType to(const UnitType& rs) {
+        return BWAPI::UnitType {rs.id};
+    }
+};
+
+template<>
+struct Cast<BWAPI::EventType::Enum> {
+    typedef EventType ToType;
+    inline static EventType from(const BWAPI::EventType::Enum& bw) {
+        return EventType { (int) bw };
+    }
+    inline static BWAPI::EventType::Enum to(const EventType& rs) {
+        return BWAPI::EventType::Enum(rs.id);
+    }
+};
+
+template<>
+struct Cast<std::string> {
+    typedef BwString* ToType;
+    inline static BwString* from(const std::string& bw) {
+        // TODO who is going to release the memory?
+        return BwString_new(bw.c_str(), (int) bw.length());
+    }
+    inline static std::string to(const BwString* rs) {
+        return std::string(BwString_data(rs));
+    }
+};
+
+template<>
+struct Cast<BWAPI::Unit> {
+    // BWAPI::Unit is already a pointer
+    typedef Unit* ToType;
+    inline static Unit* from(const BWAPI::Unit& bw) {
+        return reinterpret_cast<Unit*>(const_cast<BWAPI::Unit>(bw));
+    }
+    inline static BWAPI::Unit to(const Unit* rs) {
+        return reinterpret_cast<BWAPI::Unit>(const_cast<Unit*>(rs));
+    }
+};
+
+template<>
+struct Cast<BWAPI::Player> {
+    // BWAPI::Player is already a pointer
+    typedef Player* ToType;
+    inline static Player* from(const BWAPI::Player& bw) {
+        return reinterpret_cast<Player*>(const_cast<BWAPI::Player>(bw));
+    }
+    inline static BWAPI::Player to(const Player* rs) {
+        return reinterpret_cast<BWAPI::Player>(const_cast<Player*>(rs));
+    }
+};
+
+template<>
+struct Cast<BWAPI::Event> {
```

Comment:
BWAPI::Event is a read-only struct. There is no way to pass it as an arg to BWAPI